lms/djangoapps/lti_provider/users.py

```diff
@@ -44,34 +48,36 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
         switch_user(request, lti_user, lti_consumer)
 
 
-def create_lti_user(lti_user_id, lti_consumer):
+def create_lti_user(lti_user_id, lti_consumer, email=None):
     """
     Generate a new user on the edX platform with a random username and password,
     and associates that account with the LTI identity.
     """
-    edx_password = str(uuid.uuid4())
+    edx_user = User.objects.filter(email=email).first() if email else None
 
-    created = False
-    while not created:
-        try:
-            edx_user_id = generate_random_edx_username()
-            edx_email = f"{edx_user_id}@{settings.LTI_USER_EMAIL_DOMAIN}"
-            with transaction.atomic():
-                edx_user = User.objects.create_user(
-                    username=edx_user_id,
-                    password=edx_password,
-                    email=edx_email,
-                )
-                # A profile is required if PREVENT_CONCURRENT_LOGINS flag is set.
-                # TODO: We could populate user information from the LTI launch here,
-                # but it's not necessary for our current uses.
-                edx_user_profile = UserProfile(user=edx_user)
-                edx_user_profile.save()
-            created = True
-        except IntegrityError:
-            # The random edx_user_id wasn't unique. Since 'created' is still
-            # False, we will retry with a different random ID.
-            pass
+    if not edx_user:
+        created = False
+        edx_password = str(uuid.uuid4())
+        while not created:
+            try:
+                edx_user_id = generate_random_edx_username()
+                edx_email = f"{edx_user_id}@{settings.LTI_USER_EMAIL_DOMAIN}"
+                with transaction.atomic():
+                    edx_user = User.objects.create_user(
+                        username=edx_user_id,
+                        password=edx_password,
+                        email=edx_email,
+                    )
+                    # A profile is required if PREVENT_CONCURRENT_LOGINS flag is set.
+                    # TODO: We could populate user information from the LTI launch here,
+                    # but it's not necessary for our current uses.
+                    edx_user_profile = UserProfile(user=edx_user)
+                    edx_user_profile.save()
+                    created = True
+            except IntegrityError:
+                # The random edx_user_id wasn't unique. Since 'created' is still
+                # False, we will retry with a different random ID.
+                pass
 
     lti_user = LtiUser(
         lti_consumer=lti_consumer,
```

Comment:
@tecoholic 
I noticed that the `UNIQUE KEY `edx_user_id` (`edx_user_id`),` will fail if you have multiple LTI Consumer connections from the same email on two separate LTI Consumer instances.

Probably need to remove this unique key constraint on the database to fix this issue. This is because the `edx_user` id is found based on email.

There may not be a case for this because most likely a user won't be enrolled at two separate LMS institutions with the same email address. Let me know what you think about this.
```
CREATE TABLE `lti_provider_ltiuser` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `lti_user_id` varchar(255) NOT NULL,
  `edx_user_id` int(11) NOT NULL,
  `lti_consumer_id` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `edx_user_id` (`edx_user_id`),
  UNIQUE KEY `lti_provider_ltiuser_lti_consumer_id_lti_user_id_d684c543_uniq` (`lti_consumer_id`,`lti_user_id`),
  CONSTRAINT `lti_provider_ltiuser_edx_user_id_e06b2db5_fk_auth_user_id` FOREIGN KEY (`edx_user_id`) REFERENCES `auth_user` (`id`),
  CONSTRAINT `lti_provider_ltiuser_lti_consumer_id_c1b42ee3_fk_lti_provi` FOREIGN KEY (`lti_consumer_id`) REFERENCES `lti_provider_lticonsumer` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2011 DEFAULT CHARSET=latin1Ω
```

Reply:
@ztraboo Thanks for catching this edge case. This will be a real issue where multiple consumers are configured – for example, Canvas allows LTI Tools to be added per course (so does Open edX actually). So, there could be multiple courses configured as independent LTI Consumers in the same instance. This will lead to `LtiUser` creation failing due to this `UNIQUE` constraint. So maintaining the combination of uniqueness of `lti_consumer_id` and the `lti_user_id` should be sufficient. Expecting an `edx_user` to use only one LTI User will not work with the auto-linking. I will update the PR to handle this.

Reply:
@ztraboo Replaced the one-to-one relationship with a one-to-many in [this commit](https://github.com/openedx/edx-platform/pull/33310/commits/ca81ca562dc86ade17e45551c1a5cec8b1fb0589). Now multiple `LtiUser`s can be linked to the same `edx_user`.